psydac/api/ast/linalg.py

```diff
@@ -197,24 +203,38 @@ def _initialize(self, ndim, **kwargs):
             if ndim > 1:
                 for i,j in zip(indices1[::-1], target.args[::-1]):
                     for_body = [For(i,j, for_body)]
```

Comment:
This `LinearOperatorDot._initialize()` function is full of these if/else statements, where the 1D case is separated from the 2D+ cases. I don't really see why they are needed, as I think that the recursive construction of the for loops should work also in the 1D case.

Reply:
because we had the `product` iterator that works with 2D and 3D loops and not with 1D, so I wanted to keep the same construct of the code to put it back once we add some automatic optimization for the `product` iterator

Reply:
Sure, but we don't use `product` anymore, and if we really want to use it we can pass a single iterator to it in the 1D case: it will work just fine. Further, some loops are now merged using OpenMP pragmas. And in the future we will add even more loops by using loop tiling.

In conclusion I think that we should clean up this code and remove those `if` statements that are now useless.  There is too much code duplication right now...

Reply:
I have checked, there is an issue with the 1D case, statements like this one `variables('i1:%s'%(ndim+1),  'int')` will not give us a list of length 1 but will give us one element, and this makes treating the 1d case different from the other cases

Reply:
If this is the problem, we can simply create our own function `variables_list` that returns a list of variables in all cases.